kernel/src/drivers/terminal/mod.rs

```diff
@@ -0,0 +1,316 @@
+use core::ops::Add;
+use core::fmt;
+use core::result::Result;
+use spin::Mutex;
+use drivers::vga::{self, VgaWriter};
+use color::Color;
+
+pub static WRITER: Mutex<VgaWriter> = Mutex::new(VgaWriter::new());
+pub static STDOUT: Mutex<TextArea<VgaWriter>> = Mutex::new(TextArea::new(&WRITER, Point::new(0, 0), Point::new(vga::RESOLUTION_X, vga::RESOLUTION_Y)));
+
+/// A general error terminal error
+#[derive(Debug)]
+pub enum TerminalWriteError<T: TerminalWriter> {
+    OutputError(T::Error),
+    BackspaceUnsupported,
+    BackspaceUnavailable,
+    OutOfBounds(Point),
+}
+
+/// A terminal with a constant stream of characters.
+#[allow(dead_code)] // to be used as API
+pub struct StreamTerminal<'a, T: TerminalWriter + 'a> {
+    writer: &'a Mutex<T>,
+    color: TerminalColor,
+}
+
+#[allow(dead_code)] // to be used as API
+impl<'a, T: TerminalWriter + 'a> StreamTerminal<'a, T> {
+    pub const fn new(writer: &'a Mutex<T>) -> Self {
+        StreamTerminal {
+            writer: writer,
+            color: TerminalColor::new(Color::White, Color::Black),
+        }
+    }
+}
+
+impl<'a, T: TerminalWriter + 'a> Terminal<'a, T> for StreamTerminal<'a, T> {
+    fn set_color(&mut self, color: TerminalColor) -> Result<(), TerminalWriteError<T>> {
+        self.color = color;
+        Ok(())
+    }
+
+    fn write(&mut self, char: char) -> Result<(), TerminalWriteError<T>> {
+        let color = self.color;
+        self.write_colored(char, color)
+    }
+
+    fn write_string(&mut self, str: &str) -> Result<(), TerminalWriteError<T>> {
+        let color = self.color;
+        self.write_string_colored(str, color)
+    }
+
+    fn write_colored(&mut self, char: char, color: TerminalColor) -> Result<(), TerminalWriteError<T>> {
+        self.writer.lock().write(TerminalCharacter {
+            character: char,
+            color: color,
+        }).map_err(|e| TerminalWriteError::OutputError(e))?;
+
+        Ok(())
+    }
+
+    fn write_string_colored(&mut self, str: &str, color: TerminalColor) -> Result<(), TerminalWriteError<T>> {
+        for char in str.chars() {
+            self.write_colored(char, color)?;
+        }
+        Ok(())
+    }
+
+    fn backspace(&mut self) -> Result<(), TerminalWriteError<T>> {
+        Err(TerminalWriteError::BackspaceUnsupported)
```

Comment:
Would *all* stream terminals not support backspace?

Reply:
I don't know. Stream terminals are pretty fuzzy to me.

Reply:
Perhaps this should be decided by the underlying driver... But also some may support it at a protocol layer, complicating things. Maybe this is fine though.

Reply:
... since it's just a simple wrapper